Gh-357: Better describe FederatedOperation merge functions (#411)
gchq/gaffer-doc: docs/reference/operations-guide/federated-operation.md

## The Federated Operation

The Federated Operation is an operation which can be used against a [Federated
Store](../../administration-guide/gaffer-stores/federated-store.md). The
operation is used to send a single or a chain of operations to one or more
graphs within a federated store. It can be configured to merge results
differently depending on the [`mergeFunction`](#the-merge-function) passed to it.

### Parameters
The Federated Operation has 3 key parameters: `operation`, `graphIds` and
`mergeFunction`:

!!! example ""
    ``` json
    {
        "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
        "operation": {
            "class": "uk.gov.gchq.gaffer.operation.impl.get.GetAllElements"
        },
        "graphIds": [ "GraphA", "GraphB" ],
        "mergeFunction": {
            "class": "uk.gov.gchq.gaffer.federatedstore.util.ConcatenateMergeFunction"
        }
    }
    ```

#### Required parameter: operation

This is the Operation you wish to be federated to the subgraphs. This can be a

> single Operation or an OperationChain. If you use an OperationChain, then the

Reviewer: Might be worth a link to the Operation Chain docs now they are merged in?

For example, when GetElements is used as the operation inside a
FederatedOperation and you haven't specified a `mergeFunction`, the
function `ApplyViewToElementsFunction` will be used by default, unless an
admin has configured a different function.

See the default mergeFunctions for the operations below.

##### Default Store Configured Merge Functions

| Operation         | Merge function                                             |
|-------------------|------------------------------------------------------------|
| GetElements       | [ApplyViewToElementsFunction](#applyviewtoelementsfunction)|
| GetAllElements    | [ApplyViewToElementsFunction](#applyviewtoelementsfunction)|
| GetSchema         | [MergeSchema](#mergeschema)                                |
| GetTraits         | [CollectionIntersect](#collectionintersect)                |
| others            | [ConcatenateMergeFunction](#concatenatemergefunction)      |

### Sending Operations To Federated Stores

In these examples we do not specify the `mergeFunction` parameter. This would
therefore use the default `mergeFunction` for the specific operation.

> ??? example "Sending a single operation to one Subgraph in your Federated Store"

Reviewer: ```suggestion
??? example "Sending a single operation to one subgraph in your Federated Store"
```

## The Federated Operation

The Federated Operation is an operation which can be used against a [Federated
Store](../../administration-guide/gaffer-stores/federated-store.md). The
operation is used to send a single or a chain of operations to one or more
graphs within a federated store. It can be configured to merge results
differently depending on the [`mergeFunction`](#the-merge-function) passed to it.

### Parameters
The Federated Operation has 3 key parameters: `operation`, `graphIds` and
`mergeFunction`:

!!! example ""
    ``` json
    {
        "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
        "operation": {
            "class": "uk.gov.gchq.gaffer.operation.impl.get.GetAllElements"
        },
        "graphIds": [ "graphA", "graphB" ],
        "mergeFunction": {
            "class": "uk.gov.gchq.gaffer.federatedstore.util.ConcatenateMergeFunction"
        }
    }
    ```

#### Required parameter: operation

This is the Operation you wish to be federated to the subgraphs. This can be a
single Operation or an OperationChain. If you use an OperationChain, then the
whole chain will be sent to the subgraphs.  

#### Optional parameter: graphIds

> This is a list of graph IDs which you want to send the operation to.  

Reviewer 2: Could also mention that this can be a single graph ID, just to be clear.

If the user does not specify `graphIds` in the Operation, then the
`storeConfiguredGraphIds` for that store will be used. If the admin has not
configured the `storeConfiguredGraphIds` then all graphIds will be used. 

#### Optional parameter: mergeFunction

The `mergeFunction` parameter is the Function you want to use when merging the
results from the subgraphs.  

> If the user does not specify a `mergeFunction` then it will be selected from the

Reviewer 2: I'd suggest 2nd person here as this aimed at users and not admins.

`storeConfiguredMergeFunctions` for that store. If the admin has not configured
the `storeConfiguredMergeFunctions`, it will contain pre-populated
`mergeFunctions`. Lastly, if a suitable `mergeFunction` is not found then a

> default `ConcatenateMergeFunction` is used.  

Reviewer 2: These are "pre-populated" or "not found" from an implementation perspective. From a user perspective, all that matters is that if not specified by the user then an admin configured default is used, else the defaults in the table further down are used.

> For example, when GetElements is used as the operation inside a
> FederatedOperation and the user hasn't specified a `mergeFunction`, the
> pre-populated `ApplyViewToElementsFunction` will be selected from
> `storeConfiguredMergeFunctions`, unless the admin configured it to use something
> else.

Reviewer 2: ```suggestion
For example, when GetElements is used as the operation inside a
FederatedOperation and you haven't specified a `mergeFunction`, the
function `ApplyViewToElementsFunction` will be used by default, unless an
admin has configured a different function.
```

See the default mergeFunctions for the operations below.

> ##### Default storeConfiguredMergeFunctions

Reviewer 3: Would it be maybe better having this table under the "The Merge Function" section further down?

Author: I think it would be better here as it directly links to the sub section above.

> ##### Default storeConfiguredMergeFunctions

Reviewer 4: This comes out in capitals so should probably be spaced

| Operation         | Merge function                                             |
|-------------------|------------------------------------------------------------|
| GetElements       | [ApplyViewToElementsFunction](#applyviewtoelementsfunction)|
| GetAllElements    | [ApplyViewToElementsFunction](#applyviewtoelementsfunction)|
| GetSchema         | [MergeSchema](#mergeschema)                                |
| GetTraits         | [CollectionIntersect](#collectionintersect)                |
| others            | [ConcatenateMergeFunction](#concatenatemergefunction)      |

### Sending Operations To Federated Stores

> In these examples we do not specify the mergeFunction paramater. This would

Reviewer 3: Be consistent whether you use back ticks on `mergeFunction` or not

Also typo paramater -> parameter

                    "class": "ExampleOperation2"
                },
                "graphIds": [ "graphB" ]
            }
        ]
    }
    ```
??? example "Sending a single operation to one subgraph and a chain of operations to another subgraph in your Federated Store"
    ```json
    {
        "class": "uk.gov.gchq.gaffer.operation.OperationChain",
        "operations": [
            {
                "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
                "operation": {
                    "class": "uk.gov.gchq.gaffer.operation.OperationChain",
                    "operations": {
                        [
                            "class": "ExampleOperation1",
                            "class": "ExampleOperation2"
                        ]
                    }
                },
                "graphIds": [ "graphA" ]
            },
            {
                "class": "uk.gov.gchq.gaffer.federatedstore.operation.FederatedOperation",
                "operation": {
                    "class": "ExampleOperation3"
                },
                "graphIds": [ "graphB" ]
            }
        ]
    }
    ```

> As you can see in the example above we have wrapped an OpertationChain inside a
> Federated Operation.

Reviewer 2: ```suggestion
In the example above we've wrapped an Operation Chain inside a Federated Operation.
```

### The Merge Function

> Merge functions  dictate how the FederatedStore will merge results from

Reviewer 4: Extra space after functions

different subgraphs dependent on the Operation.

The examples below refer to this graph.

```mermaid
graph LR
    subgraph Federated Store
        subgraph Graph A
            direction TB
            A((Person<br>id: 1<br>name: bob)) --works as--> B((Job<br>id: 2<br>title: builder))
            A --lives in--> C((Place<br>id: 3<br>name: bobsville))
        end
            subgraph Graph B
            direction TB
            D((Person<br>id: 1<br>name: bob)) --drives--> E((Vehicle<br>id: 4<br>type: digger))
            E --make/model--> F((Manufacturer<br>id: 5<br>name: JCB))
        end
    end
```

#### ApplyViewToElementsFunction

This merge function uses the View from the operation and applies it to all of
the results, meaning the results are now re-aggregated and re-filtered using the
Schema, locally in the FederatedStore. This makes the results look like they
came from one graph, rather than getting back an array of Elements from the
different subgraphs.  

> ??? Example "Example using the GetAllElements operation"

Reviewer 3: Is it worth adding a simple `View` to this example to demo how the merge function would apply it to the merged results from the two graphs e.g. like 

```json
"view": {
    "entities": {
        "Person": {
            "properties": [
                "Bob"
            ]
        }
    }
}
```

Author: Added in the count property to the Person entity which will display this merging feature.

            "group": "Drives",
            "source": "1",
            "destination": "4",
            "directed": true,
            "matchedVertex": "SOURCE",
            "properties": {}
        },
        {
            "class": "uk.gov.gchq.gaffer.data.element.Entity",
            "group": "Vehicle",
            "vertex": "4",
            "properties": {
            "type": "Digger"
            }
        },
        {
            "class": "uk.gov.gchq.gaffer.data.element.Entity",
            "group": "Manufacturer",
            "vertex": "5",
            "properties": {
            "name": "JCB"
            }
        },
        {
            "class": "uk.gov.gchq.gaffer.data.element.Entity",
            "group": "Job",
            "vertex": "2",
            "properties": {
            "type": "Builder"
            }
        },
        {
            "class": "uk.gov.gchq.gaffer.data.element.Entity",
            "group": "Person",
            "vertex": "1",
            "properties": {
            "name": "Bob"
            }
        },
        {
            "class": "uk.gov.gchq.gaffer.data.element.Entity",
            "group": "Person",

>             "vertex": "1",

Reviewer 4: Might be worth describing the schema of these graphs. Specifically, the vertex `1` appears twice in the results, but most schemas would aggregate these results together. Perhaps adding a count property which gets summed would be a good way to demonstrate this. The results could be contrasted to `ConcatenateMergeFunction` which would list them as separate entities.

Author: Added this in

            "class": "java.lang.String"
            },
            "id.place.string": {
            "description": "A basic type to hold the string id of a place entity",
            "class": "java.lang.String"
            },
            "id.job.string": {
            "description": "A basic type to hold the string id of a job entity",
            "class": "java.lang.String"
            },
            "property.string": {
            "description": "A type to hold string properties of entities",
            "class": "java.lang.String"
            },
            "true": {
            "class": "java.lang.Boolean",
            "validateFunctions": [
                {
                "class": "uk.gov.gchq.koryphe.impl.predicate.IsTrue"
                }
            ]
            },
            "id.manufacturer.string": {
            "description": "A basic type to hold the string id of a manufacturer entity",
            "class": "java.lang.String"
            },
            "id.vehicle.string": {
            "description": "A basic type to hold the string id of a vehicle entity",
            "class": "java.lang.String"
            }
        }
    }
    ```

> #### ConcatenateMergeFunction

Reviewer 2: Might be good to mention that with this you can get duplicate results. As seen in the traits example query.
